src/audio/mod.rs

```diff
@@ -1,44 +1,41 @@
-use crate::{Pipeline, Result};
+use crate::{ring_buffer, Pipeline, Result};
 use anyhow::anyhow;
 use cpal::{
     traits::{DeviceTrait, HostTrait, StreamTrait},
-    Device,
+    Device, StreamConfig,
 };
-use ringbuf::{Consumer, Producer, RingBuffer};
+use ringbuf::RingBuffer;
 use std::time::Duration;
 
 const LATENCY: f32 = 20.0;
 
-pub fn run(mut pipeline: Pipeline) -> Result<()> {
-    let (input_device, output_device) = devices()?;
-
-    let config: cpal::StreamConfig = input_device.default_input_config()?.into();
-
+pub fn run(
+    input_device: &Device,
+    output_device: &Device,
+    config: &StreamConfig,
+    mut pipeline: Pipeline,
+) -> Result<()> {
     let latency_num_frames = (LATENCY / 1_000.0) * config.sample_rate.0 as f32;
     let latency_num_samples = latency_num_frames as usize * config.channels as usize;
 
     let ring = RingBuffer::new(latency_num_samples * 2);
     let (mut producer, mut consumer) = ring.split();
 
-    for _ in 0..latency_num_samples {
-        producer.push(0.0).unwrap();
-    }
+    ring_buffer::write_empty_samples(&mut producer, latency_num_samples)?;
 
     let input_data_fn = move |data: &[f32], _: &cpal::InputCallbackInfo| {
-        let output_fell_behind = write_frame(&mut producer, data);
-
-        if output_fell_behind {
-            eprintln!("output stream fell behind: try increasing latency");
+        if let Err(e) = ring_buffer::write_frame(&mut producer, data) {
+            eprintln!("input: {:?}", e);
         }
     };
 
     let output_data_fn = move |output: &mut [f32], _: &cpal::OutputCallbackInfo| {
-        let (frame, input_fell_behind) = read_frame(&mut consumer, output.len());
-        pipeline.process(&frame, output);
+        let result = ring_buffer::read_frame(&mut consumer, output.len())
+            .and_then(|frame| pipeline.process(&frame, output));
```

Comment:
`and_then()` is equivalent to `then` on JS promises, or `flatMap` on JS arrays: it's a `map` over the value of the result that returns a new `Result` and is only called if the result is `Ok` and not `Err`.

- `read_frame` produces a `Result<Vec<f32>>`
- the `and_then` closure has the type `Vec<f32> -> Result<()>`

It's pretty much equivalent to the `?` syntactic sugar, but I couldn't use the sugar in this case because I didn't want to early return on `Err`, I wanted to log instead.

The `|param| do_something(param)` syntax is a [closure](https://doc.rust-lang.org/rust-by-example/fn/closures.html).